src/group/mls_group/new_from_welcome.rs

```diff
@@ -63,43 +63,38 @@ impl MlsGroup {
         let group_info_bytes = welcome_key
             .aead_open(welcome.encrypted_group_info(), &[], &welcome_nonce)
             .map_err(|_| WelcomeError::GroupInfoDecryptionFailure)?;
-        let mut group_info = GroupInfo::decode_detached(&group_info_bytes)?;
+        let group_info = GroupInfo::decode_detached(&group_info_bytes)?;
         let path_secret_option = group_secrets.path_secret;
 
         // Build the ratchet tree
         // First check the extensions to see if the tree is in there.
-        let ratchet_tree_ext_index = group_info
+        let mut ratchet_tree_extensions = group_info
             .extensions()
             .iter()
-            .position(|e| e.extension_type() == ExtensionType::RatchetTree);
-        let ratchet_tree_extension = if let Some(i) = ratchet_tree_ext_index {
-            let extension = group_info.extensions_mut().remove(i);
-            // Throw an error if we there is another ratchet tree extension.
+            .filter(|e| e.extension_type() == ExtensionType::RatchetTree)
+            .collect::<Vec<&Box<dyn Extension>>>();
+
+        let ratchet_tree_extension = if ratchet_tree_extensions.is_empty() {
+            None
+        } else if ratchet_tree_extensions.len() == 1 {
+            let extension = ratchet_tree_extensions
+                .pop()
+                // Unwrappig here is safe because we know we only have one element
+                .unwrap()
+                .as_ratchet_tree_extension()
+                // Unwrapping here is safe, because we know the extension type already
+                .unwrap()
+                // We clone the nodes here upon extraction, so that we don't have to clone
```

Comment:
Not sure I understand why they need cloning. But this is a performance issue that will either become one later and we can fix it then, or it doesn't 🙂

Reply:
We need an actual copy here for the Welcome message.